crates/networking/rpc/engine/fork_choice.rs

```diff
@@ -157,3 +150,138 @@ impl RpcHandler for ForkChoiceUpdatedV3 {
         serde_json::to_value(response).map_err(|error| RpcErr::Internal(error.to_string()))
     }
 }
+
+trait ForkChoiceUpdatedImpl {
+    fn method(&self) -> String;
+    fn request(&self) -> &ForkChoiceUpdated;
+    fn version(&self) -> u8;
+    fn validate(
+        &self,
+        attributes: &PayloadAttributes,
+        chain_config: ChainConfig,
+        head_block: BlockHeader,
+    ) -> Result<(), RpcErr>;
+}
+
+#[derive(Debug)]
+pub struct ForkChoiceUpdatedV2(ForkChoiceUpdated);
+
+impl ForkChoiceUpdatedImpl for ForkChoiceUpdatedV2 {
+    fn method(&self) -> String {
+        "engine_forkchoiceUpdatedV2".to_string()
+    }
+
+    fn request(&self) -> &ForkChoiceUpdated {
+        &self.0
+    }
+
+    fn version(&self) -> u8 {
+        2
+    }
+
+    fn validate(
+        &self,
+        attributes: &PayloadAttributes,
+        chain_config: ChainConfig,
+        head_block: BlockHeader,
+    ) -> Result<(), RpcErr> {
+        if attributes.parent_beacon_block_root.is_some() {
+            return Err(RpcErr::InvalidPayloadAttributes(
+                "forkChoiceV2 with Beacon Root".to_string(),
+            ));
+        }
+        if !chain_config.is_shanghai_activated(attributes.timestamp) {
+            return Err(RpcErr::UnsuportedFork(
+                "forkChoiceV2 used to build pre-Shanghai payload".to_string(),
+            ));
+        }
+        if chain_config.is_cancun_activated(attributes.timestamp) {
+            return Err(RpcErr::UnsuportedFork(
+                "forkChoiceV2 used to build Cancun payload".to_string(),
+            ));
+        }
+        if attributes.timestamp <= head_block.timestamp {
+            return Err(RpcErr::InvalidPayloadAttributes(
+                "invalid timestamp".to_string(),
+            ));
+        }
+        Ok(())
+    }
+}
+
+impl TryFrom<ForkChoiceUpdatedV2> for RpcRequest {
+    type Error = String;
+
+    fn try_from(val: ForkChoiceUpdatedV2) -> Result<Self, Self::Error> {
+        ForkChoiceUpdated::try_from(&val)
+    }
+}
+
+impl RpcHandler for ForkChoiceUpdatedV2 {
+    fn parse(params: &Option<Vec<Value>>) -> Result<Self, RpcErr> {
+        Ok(Self(ForkChoiceUpdated::parse(params)?))
```

Comment:
where do you validate that params is in fact v2?

Reply:
The only difference is the parent_beacon_block_root that shouldn't be present in V2. And it is checked in v2 `validate()` [here](https://github.com/lambdaclass/ethrex/pull/1297/files#diff-4af4eb82669a3403cab38afef8fb3b766f9ded917858c00d0f9c905fde5bcb53R188)